pkg/apiserver/rest/dualwriter.go

```diff
@@ -260,3 +295,46 @@ func getName(o runtime.Object) string {
 	}
 	return accessor.GetName()
 }
+
+const dataSyncerInterval = 60 * time.Minute
+
+// StartPeriodicDataSyncer starts a background job that will execute the DataSyncer every 60 minutes
+func StartPeriodicDataSyncer(ctx context.Context, mode DualWriterMode, legacy LegacyStorage, storage Storage,
+	kind string, reg prometheus.Registerer, serverLockService ServerLockService, requestInfo *request.RequestInfo) {
+	klog.Info("Starting periodic data syncer for mode mode: ", mode)
+
+	// run in background
+	go func() {
+		// run it immediately
+		_, _ = runDataSyncer(ctx, mode, legacy, storage, kind, reg, serverLockService, requestInfo)
+
+		ticker := time.NewTicker(dataSyncerInterval)
+		for {
+			select {
+			case <-ticker.C:
+				_, _ = runDataSyncer(ctx, mode, legacy, storage, kind, reg, serverLockService, requestInfo)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+// runDataSyncer will ensure that data between legacy storage and unified storage are in sync.
+// The sync implementation depends on the DualWriter mode
+func runDataSyncer(ctx context.Context, mode DualWriterMode, legacy LegacyStorage, storage Storage,
+	kind string, reg prometheus.Registerer, serverLockService ServerLockService, requestInfo *request.RequestInfo) (bool, error) {
+	// ensure that execution takes no longer than necessary
+	const timeout = dataSyncerInterval - time.Minute
```

Comment:
In the data syncer, it looks like we call `LockExecuteAndRelease()` with a `maxInterval` of 30 mins. But, here we pass a timeout of 59 mins.

If the sync job runs for longer than 30mins, does this mean a second sync job could be created after 30 mins when the lock is released, while the first sync job is still running?

Reply:
Good catch, we should be able to use the same deadline for both the context and the lock

Reply:
you are right!

`LockExecuteAndRelease()` should be called with a `maxInterval` greater than `dataSyncerInterval`

Good catch, I will fix this!

---

For reference:
https://github.com/grafana/grafana/blob/e0b8a00e78da17d35d196732bfe5e6545419e8be/pkg/infra/serverlock/serverlock.go#L136-L141

Reply:
done!